# README

## Практическое занятие №16

### Тема: Async/await синтаксис для асинхронных операций. Обработка ошибок (try-catch-finally)

Репозитарий содержит материалы по практическому занятию №16, посвященному синтаксису async/await, и обработке ошибок.

### Цели занятия
- закрепление знаний, полученных на лекции;
- разбор практических примеров;
- практическое применение полученных знаний.

### Задачи занятия
1. Пристрелочная задача. Написать асинхронную функцию, которая будет обращаться за списком товаров по адресу `https://fakestoreapi.com/products`, и выводит результат консоль. Использовать синтаксис async/await.
2. Создать промис, который будет зарезолвлен через 3 секунды словом 'Success'. Используя синтаксис async/await, дождаться успешного выполнения промиса, и вывести результат в консоль. ВАЖНО! Слово `await` мы можем использовать только внутри `async`-функций, а следовательно, создадим для этих целей функцию `asyncFunction`, в которой будем работать с await.
3. Освежаем промисы. Написать функцию `promiseFactory`, которая будет создавать промисы. В аргументах она принимает `isSuccess`, `timeout`, `promiseResult`. Функция должна создать промис. Если `isSuccess = true`, то через `timeout` секунд промис должен быть зарезолвлен аргументом `promiseResult`. А если `isSuccess = false`, то через `timeout` секунд нужно вызвать `reject` с `promiseResult`.
4. Теперь работаем с нештатными ситуациями. Вызвать функцию `promiseFactory(false, 3, 'Something went wrong!');`. То есть, нужно имитировать нештатную ситуацию. И нам нужно, используя синтаксис async/await, обработать эту нештатную ситуацию - вывести причину в консоль. Здесь нам поможет `try/catch`.
5. Нам нужно создать 2 промиса, с использованием нашей функции `promiseFactory`, оба должны зарезолвиться успешно. Первый - через 2 секунды, числом 5, а второй - через 3 секунды, числом 7. С использованием async/await, вывести в консоль сумму результатов этих двух промисов.
6. Переписать задачу 5 так, чтобы промисы запускались параллельно. В этом нам может помочь `Promise.all`.
7. Пишем универсальную функцию отправки GET-запроса, которая будет возвращать JSON. Назовем ее `jsonRequest`. В аргументах она принимает URL, к которому надо обратиться. Она должна сделать запрос по этому адресу, затем преобразовать результат в JSON, и вернуть "готовый к употреблению" результат. Используем синтаксис async/await.
8. С помощью нашей функции, обратимся к адресу `https://fakestoreapi.com/products`. ВАЖНО: с учетом того, что `jsonRequest` объявлена как `async`, то она возвращает нам ПРОМИС, а следовательно, обрабатывать результат можем или тоже с использование async/await, или с помощью then.
9. Дорабатываем нашу функцию `jsonRequest`. Теперь, если в функцию передан пустой url, то нужно генерировать свое исключение с текстом "Please, provide the URL!".

### Полезные ссылки по занятию:
 - Русскоязычный гайд [async/await](https://learn.javascript.ru/async-await).